src/Xamarin.Android.Tools.Bytecode/ClassPath.cs

```diff
@@ -237,34 +237,44 @@ JavaDocletType GetDocletType (string path)
 		{
 			var kind = JavaDocletType.DroidDoc;
 			char [] buf = new char [500];
+			
 			string packagesHtml = Path.Combine (path, "packages.html");
 			if (File.Exists (packagesHtml) && File.ReadAllText (packagesHtml).Contains ("<body class=\"gc-documentation develop reference api "))
 				kind = JavaDocletType.DroidDoc2;
-			using (var reader = File.OpenText (Path.Combine (path, "index.html")))
-				reader.ReadBlock (buf, 0, buf.Length);
-			string rawHTML = new string (buf);
-			if (rawHTML.Contains ("Generated by javadoc (build 1.6"))
-				kind = JavaDocletType.Java6;
-			else if (rawHTML.Contains ("Generated by javadoc (version 1.7"))
-				kind = JavaDocletType.Java7;
-			else if (rawHTML.Contains ("Generated by javadoc (1.8"))
-				kind = JavaDocletType.Java8;
 
+			string indexHtml = Path.Combine (path, "index.html");
+			if (File.Exists (indexHtml)) {
+				using (var reader = File.OpenText (indexHtml))
+					reader.ReadBlock (buf, 0, buf.Length);
+				string rawHTML = new string (buf);
+				if (rawHTML.Contains ("Generated by javadoc (build 1.6"))
+					kind = JavaDocletType.Java6;
+				else if (rawHTML.Contains ("Generated by javadoc (version 1.7"))
+					kind = JavaDocletType.Java7;
+				else if (rawHTML.Contains ("Generated by javadoc (1.8"))
+					kind = JavaDocletType.Java8;
+			}
+			
 			// Check to see if it's an api.xml formatted doc
-			string rawXML = null;
-			using (var reader = File.OpenText (path)) {
-				int len = reader.ReadBlock (buf, 0, buf.Length);
-				rawXML = new string (buf, 0, len);
+			if (File.Exists (path)) {
+				string rawXML = null;
+				using (var reader = File.OpenText (path)) {
+					int len = reader.ReadBlock (buf, 0, buf.Length);
+					rawXML = new string (buf, 0, len);
+				}
+				if (rawXML.Contains ("<api>") && rawXML.Contains ("<package"))
+					kind = JavaDocletType._ApiXml;
 			}
-			if (rawXML.Contains ("<api>") && rawXML.Contains ("<package"))
-				kind = JavaDocletType._ApiXml;
 
 			return kind;
 		}
 
 		IAndroidDocScraper CreateDocScraper (string src)
 		{
-			switch (DocletType ?? GetDocletType (src)) {
+			if (!DocletType.HasValue)
+				DocletType = GetDocletType(src);
+			
+			switch (DocletType) {
```

Comment:
Not *exactly* relevant to your patch, but given that `CreateDocScraper()` can be called multiple times, for multiple paths -- once per value in `ClassPath.DocumentationPaths` -- then I don't think `DocletType` should be used *at all*. What happens if/when `ClassPath.DocumentationPaths` contains multiple *different* types: the first one "wins"?

This should probably just be:

```csharp
switch (GetDocletType (src)) {
```

Reply:
Any objection to making `GetDocletType` public then? need _some_ way to surface this code to tests.

Reply:
So, from what I can tell, the intention behind `DocletType` in the first place was to allow _classe-parse.exe_ to override the automatic detection.  This property gets set if you invoke said exe with `--docstype=..`. 

So I guess we should leave this behaviour as is (leave the `switch (DocletType ?? GetDocletType (..))`) and just make public the `GetDocletType` method.

Reply:
What's troubling about this statement is that [`class-parse` *always* sets `ClassPath.DocletType`](https://github.com/xamarin/java.interop/blob/bfb8492/tools/class-parse/Program.cs#L74). `docsType` is a `JavaDocletType`, *not* a `JavaDocletType?`, so if `class-parse --docstype` isn't used, then `ClassPath.DocletType` will be set to `JavaDocletType.DroidDoc`, which is almost certainly *wrong*.

I'm starting to think we should `[Obsolete]` the `ClassPath.DocletType` property, and make it an error to use it entirely. :-/

Reply:
On that matter, we could pretty much do away with all the various `<ItemGroup>` names currently out there and just use a `<JavaParameterNameDocsPath .. />` or something similar since `ClassPath` already does the work of checking.  In fact, the `ClassParse` task just concatenates all of these different items into a single set of doc paths: https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.Build.Tasks/Tasks/ClassParse.cs#L36-L41

So, if we obsolete `ClassPath.DocletType` we should also make it an error to specify `--docstype=` in class-parse.exe.

Reply:
This _might_ also explain some weirdness i've seen in getting the generator to parse parameter names for the android support library bindings which are **known** to be in the documentation... Seems like maybe it's just getting detected as a DroidDoc always.